File: frontend/src/lib/apis/finetune.ts
/**
 * 微调数据集管理 API 封装
 * 对接后端微调数据集路由（导出构建 / 版本列表 / 详情 / 样本预览）
 * 下载接口返回文件流，需带 Authorization 头单独 fetch，故不在此封装
 */
import { getData, postData } from '../api';

const BASE = '/api/v1/finetune/datasets';

// ===== 类型定义 =====

export type DatasetType = 'sft' | 'dpo' | 'embedding' | 'golden';

export type DatasetStatus = 'pending' | 'building' | 'completed' | 'failed';

export interface DatasetVersion {
  id: string;
  dataset_type: string;
  version: string;
  status: DatasetStatus;
  sample_count: number;
  filtered_stats: Record<string, unknown> | null;
  file_size_bytes: number | null;
  created_at: string;
  completed_at: string | null;
}

/** 详情额外携带构建参数 / 统计 / Celery 任务 ID */
export interface DatasetDetail extends DatasetVersion {
  params?: Record<string, unknown>;
  stats?: Record<string, unknown>;
  celery_task_id?: string;
}

export interface DatasetPage {
  items: DatasetVersion[];
  total: number;
}

export interface ExportParams {
  dataset_type: DatasetType;
  max_classification: string;
  days: number;
  min_rating: number;
  limit: number;
}

export interface ExportResult {
  export_id: string;
  task_id: string;
  status: string;
}

export interface PreviewResult {
  items: Record<string, unknown>[];
}

// ===== API 方法 =====

/** 提交数据集导出构建任务 */
export function exportDataset(params: ExportParams): Promise<ExportResult> {
  return postData<ExportResult>(`${BASE}/export`, params);
}

/** 获取数据集版本列表（支持类型过滤与分页） */
export function getDatasets(params?: {
  dataset_type?: string;
  page?: number;
  size?: number;
}): Promise<DatasetPage> {
  return getData<DatasetPage>(BASE, params as Record<string, string | number> | undefined);
}

/** 获取单个数据集版本详情 */
export function getDataset(id: string): Promise<DatasetDetail> {
  return getData<DatasetDetail>(`${BASE}/${id}`);
}

/** 预览数据集前 N 条样本 */
export function previewDataset(id: string, limit = 5): Promise<PreviewResult> {
  return getData<PreviewResult>(`${BASE}/${id}/preview`, { limit });
}

File: backend/app/models/finetune.py
"""
微调数据集导出模型 — 单一职责：定义微调数据集导出记录表。

每条记录对应一次数据集版本化导出（SFT/DPO/Embedding/Golden），
生命周期：pending → building → completed / failed。
构建产物为 JSONL 文件（路径与大小落库），样本级统计存 filtered_stats。
"""

import uuid
from datetime import datetime

from sqlalchemy import BigInteger, DateTime, ForeignKey, Index, Integer, String
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TimestampMixin, UUIDMixin


class DatasetExport(UUIDMixin, TimestampMixin, Base):
    """微调数据集导出记录表。"""

    __tablename__ = "finetune_dataset_exports"
    __table_args__ = (
        Index("ix_finetune_dataset_exports_tenant", "tenant_id"),
        Index("ix_finetune_dataset_exports_type", "tenant_id", "dataset_type"),
    )

    tenant_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), nullable=True, comment="租户 ID（多租户隔离）"
    )
    dataset_type: Mapped[str] = mapped_column(
        String(20), nullable=False, comment="数据集类型: sft/dpo/embedding/golden"
    )
    version: Mapped[str] = mapped_column(
        String(40), nullable=False, comment="版本号: v{YYYYMMDD-HHmmss}"
    )
    status: Mapped[str] = mapped_column(
        String(20), default="pending", comment="状态: pending/building/completed/failed"
    )
    sample_count: Mapped[int] = mapped_column(
        Integer, default=0, comment="导出样本数"
    )
    filtered_stats: Mapped[dict] = mapped_column(
        JSONB, default=dict, comment="过滤统计（按原因分类: classification/duplicate/...）"
    )
    file_path: Mapped[str | None] = mapped_column(
        String(500), nullable=True, comment="JSONL 文件路径"
    )
    file_size_bytes: Mapped[int] = mapped_column(
        BigInteger, default=0, comment="文件大小（字节）"
    )
    celery_task_id: Mapped[str | None] = mapped_column(
        String(64), nullable=True, comment="构建任务 Celery ID"
    )
    created_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True, comment="创建者 ID"
    )
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, comment="完成时间"
    )
    params: Mapped[dict] = mapped_column(
        JSONB, default=dict, comment="构建参数（max_classification/days/min_rating/limit）"
    )
